Server-side power API #179

ctoomey opened the pull request:
Here's a draft, scala-only implementation of both #179 and #420.

In AkkaGrpcPlugin.scala I added a new setting `akkaGrpcGeneratorOptions` with 2 options: `ServerPowerApis` to generate server power APIs, and `UsePlayActions` to generate a handler and router that implement Play's `RequestHeader => EssentialAction` instead of `HttpRequest => Future[HttpResponse]`.

The server power APIs are defined by a power API trait that extends the base trait and adds a `MetadataMap` parameter to each method.

If this looks good I’ll work on the rest of the changes (Java implementation, etc.).

johanandren commented:
I think it would be easier to review and understand what is going on if the PRs were split up and there was one with the power-user API and one with the Play-nativeness changes, you think you could do like that instead @ctoomey ? (Would be fine to base one PR branch off the other if you have changes that both need. )

raboof commented:
OK TO TEST

ctoomey commented on sbt-plugin/src/main/scala/akka/grpc/sbt/AkkaGrpcPlugin.scala:
```diff
@@ -36,18 +37,28 @@ object AkkaGrpcPlugin extends AutoPlugin {
     val akkaGrpcGeneratedSources = settingKey[Seq[AkkaGrpc.GeneratedSource]](
-      "Which of the sources to generate in addition to the gRPC protobuf messages (AkkaGrpc.Server, AkkaGrpc.Client)")
+      "Which of the sources to generate in addition to the gRPC protobuf messages (AkkaGrpc.Server, AkkaGrpc.Client, AkkaGrpc.PlayServer, AkkaGrpc.PlayClient)")
+    val akkaGrpcGeneratorOptions = settingKey[Seq[AkkaGrpc.GeneratorOption]]("Generator options (AkkaGrpc.ServerPowerApis). Empty by default")
```
Just getting some time to work on this again. I'll split up the PRs and look into your suggestions for deduping and moving the sbt config. settings.

I purposely deferred any work on Java and non-sbt builds as it's pointless until there's an agreeable Scala / sbt solution, plus I may not have sufficient experience to do those parts.

ctoomey replied:
I remember now that the problem with adding to `akkaGrpcCodeGeneratorSettings` is that those are mapped to `scalapb.compiler.GeneratorParams` and I'm pretty sure adding Akka gRPC-specific params to ScalaPB isn't a viable option.

raboof replied:
Indeed deferring work on Java and build systems other than sbt until after we're happy with the general approach makes sense.

ctoomey replied:
> It is true that different generators require different parameters, but they also overlap: for example, when passing `flat_package` to the ScalaPB generator, we also need to pass it to the other generators to make sure it refers to the correct ScalaPB-generated class names.
> 
> For this reason it seems convenient to me to present one `akkaGrpcCodeGeneratorSettings` to the user, and use a filter to only pass the 'relevant' ones to each generator. Does that make sense?

It makes sense as a goal, I'll see how difficult it looks to accomplish. I think it's not that bad to have a separate setting if it's a lot of work to get there though.

ctoomey commented on codegen/src/main/twirl/templates/PlayScala/HandlerUsingActions.scala.txt:
```diff
@@ -0,0 +1,221 @@
+      import @{service.name}.Serializers._
+
+      def handle(request: RequestHeader, method: String): EssentialAction = method match {
+        @for(method <- service.methods) {
```
I'll investigate this, but given the Akka gRPC handler returns completed HttpResponses whereas the Play one returns EssentialActions that need further processing including from filters before they generate a response, not sure wrapping is gonna work.

raboof replied:
Right, so the wrapping would take the `HttpResponse` generated by the generic handler, and convert it to an `EssentialAction` to be passed through further filters. Could that work?

ctoomey replied:
Yep it looks like something along these lines is doable.

ctoomey commented on codegen/src/main/scala/akka/grpc/gen/scaladsl/play/PlayScalaServerCodeGenerator.scala:
```diff
@@ -5,21 +5,47 @@
+  private val generateRouterUsingActions: (Logger, Service) => CodeGeneratorResponse.File = (logger, service) => {
+    val b = CodeGeneratorResponse.File.newBuilder()
+    b.setContent(RouterUsingActions(service, powerApis).body)
+    b.setName(s"${service.packageDir}/Abstract${service.name}RouterUsingActions.scala")
```
Sure, it'd be less work, but as of now the Action router takes an extra constructor parameter.

ctoomey commented on runtime/src/main/scala/akka/grpc/scaladsl/scaladsl.scala:
```diff
@@ -0,0 +1,35 @@
+import scala.collection.immutable
+
+package object scaladsl {
+  type MetadataMap = immutable.Map[String, Seq[MetadataEntry]]
```
The existing `Metadata` facade is a super limited interface that is inadequate for "real" use by service implementers. It only provides `getText()` and `getBinary()` methods returning optional values if given keys exist. It doesn't provide a means for getting all the keys nor for getting multi-valued keys. So some kind of richer interface is needed, and per discussion on #179 I kept the new-code surface small by using a `Map` and reusing `MetadataEntry`. If you want the latter to be kept internal, how would you suggest providing this richer API?

raboof replied:
Instead of exposing `MetadataEntry` we could also consider adding additional methods to the `Metadata` API. I don't have a strong preference at this point, though.

johanandren replied:
Keeping a simple facade gives us more freedom to change implementation though and optimize for performance etc. I think having the facade makes sense, having an explicit `Metadata.toSeq` to get the collection capabilities for example would perhaps allow for the more advanced use cases?

raboof commented on codegen/src/main/twirl/templates/ScalaServer/PowerApiTrait.scala.txt:
```diff
@@ -0,0 +1,27 @@
+
+import io.grpc.Status
+
+trait @{service.name}PowerApi extends @{service.name} {
```
In the future we'd want to allow finer-grained configuration to determine which methods should have the power API and which methods have the simple API. In that case perhaps `PowerApi` is not a great name - though I'm not sure about alternatives. This is OK for now.

ctoomey commented:
I've updated `AkkaGrpcPlugin` to include the additional server-side code generation options into `akkaGrpcCodeGeneratorSettings` as requested.

ctoomey commented:
@raboof @johanandren: I've made the requested changes so am ready to look at supporting java and the other build tools.

Please confirm you're good with this moving forward.

johanandren commented on runtime/src/main/scala/akka/grpc/internal/PlayRouterUsingActions.scala:
```diff
@@ -0,0 +1,54 @@
+          s"Attempted to prefix with [$prefix], yet already default prefix known to be [${this.prefix}]. " +
+          s"When binding gRPC routers the path in `routes` MUST BE `/`.")
+
+}
```
Is this still something that belongs in the other PR?

ctoomey replied:
Yep, removed it.

johanandren commented on codegen/src/main/twirl/templates/ScalaServer/Handler.scala.txt:
```diff
@@ -2,86 +2,94 @@
+        @for(method <- service.methods) {
+        case "@method.grpcName" =>
+          val responseCodec = Codecs.negotiate(request)
+          @{if(powerApis) { "val metadataMap = MetadataMap(request.headers)" } else { "" }}
```
Now this is eager for all power APIs even if they don't actually read the metadata. An actual type inbetween could lessen that to just one object allocation, and then parse the headers into our datatypes only if they are accessed.

ctoomey replied:
Reverted to using an enhanced version of the `Metadata` trait, see `runtime/src/main/scala/akka/grpc/scaladsl/Metadata.scala`.

ctoomey commented on codegen/src/main/scala/akka/grpc/gen/scaladsl/ScalaServerCodeGenerator.scala:
```diff
@@ -8,18 +8,30 @@ import akka.grpc.gen.Logger
-    b.setContent(Handler(service).body)
-    b.setName(s"${service.packageDir}/${service.name}Handler.scala")
+    b.setContent(PowerApiTrait(service).body)
+    b.setName(s"${service.packageDir}/${service.name}PowerApi.scala")
```
The base API is needed by the generated clients too.

johanandren replied:
Ah yes, didn't think of that. Ok, let keep that as is then (I still don't like it but can't come up with anything better)

johanandren commented on runtime/src/main/scala/akka/grpc/scaladsl/Metadata.scala:
```diff
@@ -18,9 +23,45 @@ import akka.util.ByteString
+    entries
+  }
+
+  override def getText(key: String): Option[String] = headers.find(_.name == key).map(_.value)
```
I guess these may be from the original implementation, however `collectFirst { case header if header.name == key => header.value }` would avoid an extra option and function allocation here

ctoomey replied:
Fixed

johanandren commented on runtime/src/main/scala/akka/grpc/scaladsl/scaladsl.scala:
```diff
@@ -1,35 +1,5 @@
-      entries
-    }
-  }
+  type MetadataMap = Map[String, List[MetadataEntry]]
```
I don't think we should have a type alias as public API, and the signature isn't that messy, so let's just use that instead.

ctoomey replied:
Fixed

johanandren commented on runtime/src/main/scala/akka/grpc/scaladsl/Metadata.scala:
```diff
@@ -18,9 +23,45 @@ import akka.util.ByteString
+        entries += (key -> (entry :: Nil))
+    }
+    entries
+  }
```
I wish we could do this with the mutable `Map.newBuilder` instead of creating a new map for each header entry. Not sure that is doable though.

ctoomey replied:
Ya, not with having to group by header name.

johanandren replied:
That wouldn't help because we'd still pass a map along and update, so it would just be a different way of doing the same.

johanandren commented on runtime/src/main/scala/akka/grpc/scaladsl/Metadata.scala:
```diff
@@ -18,9 +23,45 @@ import akka.util.ByteString
+}
+
+class MetadataImpl(headers: immutable.Seq[HttpHeader] = immutable.Seq.empty) extends Metadata {
+  lazy private val b64Decoder = Base64.getDecoder
```
This one is just returning a static final field in `java.util.Encoder` no need to introduce a lazy val for that.

ctoomey replied:
Fixed

johanandren commented:
I think this direction is good, and that you can go ahead with the Java APIs. One concern is that we have some other work coming in #449 which will likely give you a bit of conflict when that is merged, not sure how messy that will be though.

dwijnand commented:
😅 looks like no merge conflicts.

ctoomey commented:
@raboof , I'll look at addressing your comments this weekend. Re: adding an example, do you mean an example of a "power API" service method implementation illustrating using the metadata param?

ctoomey commented on codegen/src/main/scala/akka/grpc/gen/scaladsl/ScalaBothCodeGenerator.scala:
```diff
@@ -4,6 +4,6 @@
 package akka.grpc.gen.scaladsl
 
-object ScalaBothCodeGenerator extends ScalaServerCodeGenerator with ScalaClientCodeGenerator {
+object ScalaBothCodeGenerator extends ScalaServerCodeGenerator(powerApis = false) with ScalaClientCodeGenerator {
```
I actually implemented this change and then was reminded when I tried to use the updated code on a project that we can't pass through to scalaPB our additional code generator params without also modifying it to handle them -- otherwise it throws an "unrecognized parameter" exception. That's why I didn't do it that way originally.

ctoomey commented on codegen/src/main/twirl/templates/PlayScala/AkkaGrpcClientModule.scala.txt:
```diff
@@ -5,7 +5,7 @@
 
 @akka.grpc.gen.Constants.DoNotEditComment
-package @{packageName}
+@{if (packageName.nonEmpty) s"package $packageName" else ""}
```
Ok will separate it.

ctoomey replied:
Without this my test project compilation breaks so I'm going to leave this in this work in progress branch. Once this work gets closer to being done I'll do a separate PR if I think of it.

raboof replied:
OK, fine to leave it here

ctoomey commented on codegen/src/main/twirl/templates/ScalaServer/Handler.scala.txt:
```diff
@@ -2,86 +2,94 @@
 
-object @{service.name}Handler {
-  private val notFound = Future.successful(HttpResponse(StatusCodes.NotFound))
+@defining(if (powerApis) service.name + "PowerApi" else service.name) { serviceName =>
```
We need it to reference either the base service trait or the power API trait. But the handler name is independent of power APIs and so doesn't have PowerApi in it.

ctoomey commented:
Here's an example of a power API method that uses the metadata to allow only authenticated requests (indicated by a valid `Authorization` metadata value):

```
...

class GreeterServiceImpl(materializer: Materializer) extends GreeterServicePowerApi {
  import materializer.executionContext
  private implicit val mat: Materializer = materializer

  override def sayHello(in: HelloRequest, metadata: Metadata): Future[HelloReply] = {
    checkAuthentication(metadata)
    println(s"sayHello to ${in.name}")
    Future.successful(HelloReply(s"Hello, ${in.name}"))
  }

...

  def checkAuthentication(metadata: Metadata): Unit = {
    metadata.getText("authorization").map(authValue => {
      val valid = ...
      if (valid) ()
      else throw new GrpcServiceException(Status.Code.UNAUTHENTICATED)
    })
  }
}
```

raboof commented:
I like the example, I think it would be good to have it in the repository somewhere (in the tests?) to make sure it compiles and so it will be easy to include it in the documentation eventually.

I think it would be a little better still if we did not only check the authentication, but actually used something from it in the reply: this current example could also be achieved by having some 'middleware' checking the authentication and not using a Power API for the service implementations.

ctoomey commented on sbt-plugin/src/main/scala/akka/grpc/sbt/AkkaGrpcPlugin.scala:
```diff
@@ -117,41 +136,35 @@ object AkkaGrpcPlugin extends AutoPlugin {
+      case (_, Java) => logger.error("Livongo: java code generators not yet available"); Seq.empty
+      case (Client, Scala) => Seq(ScalaGenerator, toGenerator(ScalaClientCodeGenerator, scalaBinaryVersion, logger))
+      case (PlayClient, Scala) => Seq(ScalaGenerator, toGenerator(PlayScalaClientCodeGenerator, scalaBinaryVersion, logger))
+      case (Server, Scala) => Seq(ScalaGenerator, toGenerator(ScalaServerCodeGenerator(serverPowerApis), scalaBinaryVersion, logger))
```
You're right, that was broken. But when I looked at how to update this logic to handle using that correctly with all the possible combinations (Client, PlayClient) x (Server, PlayServer) it looked to be quite ugly. A cleaner approach I implemented was creating a standalone `ScalaTraitCodeGenerator` to generate the service trait, and changing the client and server code generators to only generate their files and not also the trait. Doing so eliminated the need for `ScalaBothCodeGenerator` and allowed this logic in `AkkaGrpcPlugin` to stay simple.

raboof replied:
I agree the current approach there is relatively ugly, and it would be worth investigating how it can be improved. I was a bit afraid that doing this in this PR might make it bigger and slow it down, which is why I didn't propose doing it here - if you don't mind then doing it here is OK by me.

ctoomey commented:
> I like the example, I think it would be good to have it in the repository somewhere (in the tests?) to make sure it compiles and so it will be easy to include it in the documentation eventually.
> 
> I think it would be a little better still if we did not only check the authentication, but actually used something from it in the reply: this current example could also be achieved by having some 'middleware' checking the authentication and not using a Power API for the service implementations.

I'll add this example enhanced as you suggest in a test once we've stopped iterating on the Scala solution I've implemented, which I hope is soon.

raboof commented on sbt-plugin/src/main/scala/akka/grpc/sbt/AkkaGrpcPlugin.scala:
```diff
@@ -117,41 +137,39 @@ object AkkaGrpcPlugin extends AutoPlugin {
-        }
-    }
+    lazy val serverPowerApis = options.contains(GeneratorOption.ServerPowerApis.setting)
+    lazy val baseGenerators: Seq[Generator] = Seq(ScalaGenerator, toGenerator(ScalaTraitCodeGenerator, scalaBinaryVersion, logger))
```
Having these separate seems fine by me. It's hard to say whether this will introduce complexity in the Maven and Gradle plug-ins, but we'll find out when those are enabled again.

raboof commented on sbt-plugin/src/main/scala/akka/grpc/sbt/AkkaGrpcPlugin.scala:
```diff
@@ -32,25 +35,41 @@ object AkkaGrpcPlugin extends AutoPlugin {
+
+      case object Client extends GeneratedClient
+      case object Server extends GeneratedServer
+      case object PlayClient extends GeneratedClient
```
We are in the process of moving some Play-specific parts to its own repo (http://github.com/playframework/play-grpc, #474). That might introduce some conflicts here. I don't think we can really prevent that, so we'll just have to fix those as they come up.

ctoomey commented:
@raboof, I've started working on porting my changes to generated Java code, but it's been a long time since I've used Java so will need some help.

First off, here's what I'm guessing the `akka.grpc.javadsl.Metadata.asMap` method interface should look like -- what do you think?

```
import java.lang.{Iterable => jIterable}
import java.util.{Optional, Map => jMap}
...
trait Metadata {
  ...
  def asMap: jMap[String, jIterable[MetadataEntry]]
}
```

Second, I don't know how to model/implement the Java equivalent of `akka.grpc.scaladsl.MetadataEntry`. Since it's so trivial can you just suggest the Java code here for it?

raboof commented:
I agree from a Java API perspective exposing the `MetadataEntry` hierarchy gets quite complicated. We have some similar situations in Akka HTTP, but I'm not sure it is worth the complexity here. Do we really need the `asMap` already? If not we could perhaps remove it and leave it for another PR.

ctoomey commented:
Note: I've finished the Java implementation of the power APIs and the maven plugin update, so AFAIK the only things left are the gradle plugin and updating the doc.

johanandren commented:
I haven't done a thorough review (traveling) but just wanted to say that this is shaping up nicely!

ctoomey commented:
Hi all, I've added documentation for the server power APIs and that completes everything that was left to do AFAIK. Please review so we can finally get it merged and I can move on to similarly finishing #447 .

ctoomey commented on codegen/src/main/scala/akka/grpc/gen/Main.scala:
```diff
@@ -42,71 +39,65 @@ object Main extends App {
-
-  System.out.write(out.toByteArray)
-  System.out.flush()
+    //    throw new RuntimeException(s"just before executing code generators")
```
done

ctoomey commented on codegen/src/main/scala/akka/grpc/gen/Main.scala:
```diff
@@ -42,71 +39,65 @@ object Main extends App {
-  object CombinedPlayScalaBothCodeGenerator extends PlayScalaClientCodeGenerator with ScalaClientCodeGenerator with PlayScalaServerCodeGenerator with ScalaServerCodeGenerator {
-    override def name = "combined-play-scala-both"
+    codeGenerators.foreach { g =>
+      //      println("executing generator " + g.name)
```
done

ctoomey commented on codegen/src/main/scala/akka/grpc/gen/scaladsl/ScalaTraitCodeGenerator.scala:
```diff
@@ -0,0 +1,23 @@
+import com.google.protobuf.compiler.PluginProtos.CodeGeneratorResponse
+import templates.ScalaCommon.txt.ApiTrait
+
+object ScalaTraitCodeGenerator extends ScalaCodeGenerator {
```
Having "service" in the name would be inconsistent w/ all the other code generators, none of which do. And I figured "trait" was more apropos for Scala than "interface".

But if you feel strongly it should be `ServiceInterfaceCodeGenerator` or another variant, let me know and I'll rename it.

johanandren replied:
Nah, I don't feel strongly about it, let's keep as is.

ctoomey commented on codegen/src/main/scala/akka/grpc/gen/scaladsl/play/PlayScalaServerCodeGenerator.scala:
```diff
@@ -5,21 +5,23 @@
-  override def perServiceContent = super.perServiceContent + generateRouter
+  override def perServiceContent = super.perServiceContent + generateRouter() ++ (
+    if (powerApis) Set(generateRouter(powerApis))
+    else Set.empty
```
I generated both variants of the handler so that the plugin-tester tests could test both variants, so I figured we should be consistent and similarly generate both variants of the router.

But if you still think it preferable to only generate the power API router let me know and I'll make the change.

johanandren replied:
Ah ok, let's keep it like this then

ctoomey commented on docs/src/main/paradox/server/walkthrough.md:
```diff
@@ -256,3 +256,28 @@ Java
+and @ref[Maven](../buildtools/maven.md) for how to set this build option. Note that this option doesn't effect the
+generated client stubs.
+
+Here's what the generated server power APIs look like:
```
done

johanandren commented:
We'll want a final LGTM from @raboof as well but he is on vacation right now.

ctoomey commented:
Thanks @johanandren, I hope @raboof comes back from vacation in a merging mood :-).

raboof commented on codegen/src/main/scala/akka/grpc/gen/Main.scala:
```diff
@@ -7,17 +7,14 @@ package akka.grpc.gen
 
 // This is the protoc plugin that the gradle plugin uses
 object Main extends App {
+  //  throw new RuntimeException(s"beginning of Main")
```
#528

ctoomey commented:
Yay, thanks @raboof! Happy to be able to contribute back our enhancements to this important project.
